rust/arrow/src/array/array_binary.rs

```diff
@@ -364,62 +365,135 @@ impl FixedSizeBinaryArray {
-
-        let num_bytes = bit_util::ceil(len, 8);
-        let mut null_buf = MutableBuffer::from_len_zeroed(num_bytes);
-        let null_slice = null_buf.as_slice_mut();
-
-        data.iter().enumerate().for_each(|(i, entry)| {
-            if entry.is_some() {
-                bit_util::set_bit(null_slice, i);
+            Ok(())
+        })?;
+
+        if len == 0 {
+            return Err(ArrowError::InvalidArgumentError(
+                "Input iterable argument has no data".to_owned(),
+            ));
+        }
+
+        if prepend > 0 {
+            let extend_size = size * prepend;
+            let copy_size = buffer.len();
+            buffer.resize(copy_size + extend_size, 0);
+            unsafe {
+                let src = buffer.as_ptr();
+                let dst = buffer.as_mut_ptr().add(extend_size);
+                std::ptr::copy(src, dst, copy_size);
+                buffer.as_mut_ptr().write_bytes(0, extend_size);
             }
-        });
+        }
 
-        let data = data
-            .into_iter()
-            .flat_map(|e| e.unwrap_or_else(|| vec![0; size]))
-            .collect::<Vec<_>>();
-        let data = ArrayData::new(
+        let array_data = ArrayData::new(
             DataType::FixedSizeBinary(size as i32),
             len,
             None,
             Some(null_buf.into()),
             0,
-            vec![Buffer::from(&data)],
+            vec![buffer.into()],
             vec![],
         );
-        FixedSizeBinaryArray::from(Arc::new(data))
+        Ok(FixedSizeBinaryArray::from(Arc::new(array_data)))
+    }
+
+    /// Create an array from an iterable argument of byte slices.
+    ///
+    /// # Errors
+    ///
+    /// Returns error if argument has length zero, or sizes of nested slices don't match.
+    pub fn try_from_iter<T, U>(mut iter: T) -> Result<Self, ArrowError>
+    where
+        T: Iterator<Item = U>,
+        U: AsRef<[u8]>,
+    {
+        let mut len = 0;
+        let mut size = 0;
+        let mut buffer = MutableBuffer::from_len_zeroed(0);
+        iter.try_for_each(|item| -> Result<(), ArrowError> {
+            let slice = item.as_ref();
+            if size != slice.len() {
+                if size == 0 {
+                    size = slice.len();
+                } else {
+                    return Err(ArrowError::InvalidArgumentError(format!(
+                        "Nested array size mismatch: one is {}, and the other is {}",
+                        size,
+                        slice.len()
+                    )));
+                }
+            }
+            buffer.extend_from_slice(slice);
+
+            len += 1;
+
+            Ok(())
+        })?;
+
+        if len == 0 {
+            return Err(ArrowError::InvalidArgumentError(
+                "Input iterable argument has no data".to_owned(),
+            ));
+        }
+
+        let array_data = ArrayData::builder(DataType::FixedSizeBinary(size as i32))
+            .len(len)
```

Comment:
Consider an iterator of size `4` whose every element is `Some(&[0, 0, 0])`.

I _think_ that `len` is being increased by 1 on every group of size `size`. In the example above, won't `len` equal to `4` instead of `12`?

I _think_ that this should be `len * size`

Reply:
Yes, it will be equal to 4. But in the previous implementation, for example `impl From<Vec<Vec<u8>>> for FixedSizeBinaryArray`, when we pass for example a vec with 4 items to `from` then it will put 4 as len in the array builder. Isn't it the same there?

Reply:
You are right (always learning something new here). In FixedSizeBinary[3] , each item is `[a, b, c]`, pretty much like in an ListArray, and thus the len should be 4, not 12. So, the invariant is actually the other way around:

```rust
array.buffer()[0].len() % size == 0
array.len() * size == array.buffer()[0].len()
array.validity().len() == array.len()
```

Sorry about the noise and thank you for your patience 🥇

Reply:
Great work you do 👍